packages/cdktf-cli/lib/dependencies/package-manager.ts

```diff
@@ -0,0 +1,146 @@
+import { Language } from "@cdktf/provider-generator";
+import { existsSync } from "fs-extra";
+import path from "path";
+import { exec } from "../util";
+import { xml2js, js2xml, Element } from "xml-js";
+import { Errors } from "../errors";
+import * as fs from "fs-extra";
+
+/**
+ * manages installing, updating, and removing dependencies
+ * in the package system used by the target language of a CDKTF
+ * project
+ */
+export abstract class PackageManager {
+  constructor(protected readonly workingDirectory: string) {}
+
+  public static forLanguage(
+    language: Language,
+    workingDirectory: string
+  ): PackageManager {
+    switch (language) {
+      case Language.GO:
+        throw new Error("Go is not supported yet");
+      case Language.TYPESCRIPT:
+        return new NodePackageManager(workingDirectory);
+      case Language.PYTHON:
+        return new PythonPackageManager(workingDirectory);
+      case Language.CSHARP:
+        return new NugetPackageManager(workingDirectory);
+      case Language.JAVA:
+        return new MavenPackageManager(workingDirectory);
```

Comment:
We should setup this up in a way that it is easy to switch out a different package manager. For example, Java has multiple available package managers. Maven is one of the most common and what we support elsewhere, but Gradle is also very common.

I believe a similar situation exists for python and we already support multiple methods with the templates.

Reply:
You're right. However, currently we only support the package managers, that our template supports. However, for example, the "NodePackageManager" (poor wording) does check for a `yarn.lock` to switch between yarn and npm.
But we could split that into two separate classes in the future and introduce an abstract "isProjectUsingThisPackageManager" method that those classes also implement 😁 But I see that as out of scope for the first iteration and I think the current design is easy enough to be refactored in such a way that we don't need to adjust them right now to enable to this kind of change in the future.